File: yntra-ios/YntraIOS/ViewModels/AuthViewModel.swift
import SwiftUI
import yntra_core

@MainActor
class AuthViewModel: ObservableObject {
    @Published var isLoggedIn: Bool = false
    @Published var authRegion: String = "sv" // "sv" | "da" | "no" | "fi" | "en"
    @Published var bankIdFlowState: String = "idle" // "idle" | "qr_scan" | "verifying" | "success" | "error"
    @Published var qrData: String? = nil
    @Published var errorMessage: String? = nil

    private var activeSessionId: String? = nil
    private var isPolling = false

    func setRegion(_ region: String) {
        self.authRegion = region
    }

    func initiateBankIdLogin(provider: String) {
        self.errorMessage = nil
        self.bankIdFlowState = "connecting"
        
        Task {
            do {
                let session = try await initiateBankIdAuth(targetRole: "assistant", provider: provider)
                self.activeSessionId = session.id
                self.bankIdFlowState = session.status
                self.qrData = session.qrData
                self.startPollingStatus(sessionId: session.id)
            } catch {
                self.bankIdFlowState = "error"
                self.errorMessage = error.localizedDescription
            }
        }
    }

    private func startPollingStatus(sessionId: String) {
        guard !isPolling else { return }
        isPolling = true
        
        Task {
            while isPolling && self.bankIdFlowState != "success" && self.bankIdFlowState != "error" {
                try? await Task.sleep(nanoseconds: 2_000_000_000) // Sleep 2 seconds
                do {
                    if let session = try await getBankidAuthSession(sessionId: sessionId) {
                        self.bankIdFlowState = session.status
                        self.qrData = session.qrData
                        if session.status == "success" || session.status == "authenticated" {
                            self.isLoggedIn = true
                            self.isPolling = false
                        } else if session.status == "failed" || session.status == "error" {
                            self.errorMessage = "Authentication failed"
                            self.bankIdFlowState = "error"
                            self.isPolling = false
                        }
                    } else {
                        self.bankIdFlowState = "error"
                        self.errorMessage = "Session not found"
                        self.isPolling = false
                    }
                } catch {
                    self.bankIdFlowState = "error"
                    self.errorMessage = error.localizedDescription
                    self.isPolling = false
                }
            }
            self.isPolling = false
        }
    }

    func passwordLogin(email: String, pin: String) {
        self.errorMessage = nil
        Task {
            do {
                if let _ = try await verifyEmailPassword(email: email, password: pin) {
                    self.isLoggedIn = true
                } else {
                    self.errorMessage = "Invalid credentials"
                }
            } catch {
                self.errorMessage = error.localizedDescription
            }
        }
    }

    func logout() {
        self.isLoggedIn = false
        self.bankIdFlowState = "idle"
        self.qrData = nil
        self.activeSessionId = nil
        self.isPolling = false
    }
}

File: yntra-android/app/src/main/java/com/yntra/app/viewmodels/AuthViewModel.kt
package com.yntra.app.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import uniffi.yntra_core.*

class AuthViewModel : ViewModel() {
    private val _isLoggedIn = MutableStateFlow(false)
    val isLoggedIn: StateFlow<Boolean> = _isLoggedIn

    private val _authRegion = MutableStateFlow("sv") // "sv" | "da" | "no" | "fi" | "en"
    val authRegion: StateFlow<String> = _authRegion

    private val _bankIdFlowState = MutableStateFlow("idle") // "idle" | "qr_scan" | "verifying" | "success" | "error"
    val bankIdFlowState: StateFlow<String> = _bankIdFlowState

    private val _qrData = MutableStateFlow<String?>(null)
    val qrData: StateFlow<String?> = _qrData

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage

    private var activeSessionId: String? = null
    private var isPolling = false

    fun setRegion(region: String) {
        _authRegion.value = region
    }

    fun initiateBankIdLogin(provider: String) {
        viewModelScope.launch {
            _errorMessage.value = null
            _bankIdFlowState.value = "connecting"
            try {
                // target_role: employee / assistant
                val session = initiateBankIdAuth("assistant", provider)
                activeSessionId = session.id
                _bankIdFlowState.value = session.status // e.g. "qr_scan"
                _qrData.value = session.qrData
                startPollingStatus(session.id)
            } catch (e: Exception) {
                _bankIdFlowState.value = "error"
                _errorMessage.value = e.message
            }
        }
    }

    private fun startPollingStatus(sessionId: String) {
        if (isPolling) return
        isPolling = true
        viewModelScope.launch {
            while (isPolling && _bankIdFlowState.value != "success" && _bankIdFlowState.value != "error") {
                delay(2000)
                try {
                    val sessionOpt = getBankidAuthSession(sessionId)
                    if (sessionOpt != null) {
                        _bankIdFlowState.value = sessionOpt.status
                        _qrData.value = sessionOpt.qrData
                         if (sessionOpt.status == "success" || sessionOpt.status == "authenticated") {
                            sessionOpt.authenticatedUserId?.let { uid ->
                                com.yntra.app.SessionManager.activeUserId = uid
                            }
                            _isLoggedIn.value = true
                            isPolling = false
                        } else if (sessionOpt.status == "failed" || sessionOpt.status == "error") {
                            _errorMessage.value = "Authentication failed"
                            _bankIdFlowState.value = "error"
                            isPolling = false
                        }
                    } else {
                        _bankIdFlowState.value = "error"
                        _errorMessage.value = "Session not found"
                        isPolling = false
                    }
                } catch (e: Exception) {
                    _bankIdFlowState.value = "error"
                    _errorMessage.value = e.message
                    isPolling = false
                }
            }
            isPolling = false
        }
    }

    fun passwordLogin(email: String, pin: String) {
        viewModelScope.launch {
            _errorMessage.value = null
            try {
                val user = verifyEmailPassword(email, pin)
                if (user != null) {
                    com.yntra.app.SessionManager.activeUserId = user.id
                    user.workspaceId?.let { ws ->
                        com.yntra.app.SessionManager.activeWorkspaceId = ws
                    }
                    _isLoggedIn.value = true
                } else {
                    _errorMessage.value = "Invalid email or credentials"
                }
            } catch (e: Exception) {
                _errorMessage.value = e.message
            }
        }
    }

    fun logout() {
        _isLoggedIn.value = false
        _bankIdFlowState.value = "idle"
        _qrData.value = null
        activeSessionId = null
        isPolling = false
    }
}
